### PR TITLE
Implement `CardinalityAwareRowConverter` while doing streaming merge

### LINKED ISSUE #7200: RowConverter keeps growing in size while merging streams on high-cardinality dictionary fields
### Describe the bug

On executing `SortPreservingMerge` on multiple streams of record batches and using a high-cardinality dictionary field as the sort key, the `RowConverter` instance used to merge the multiple `RowCursorStreams` keeps growing in memory (as it keeps accumulating the dict mappings internally in the `OrderPreservingInterner` structure). This unbounded memory growth eventually causes data fusion to get killed by the OOM killer. 

### To Reproduce

Detailed steps to reproduce this issue is given [here](https://github.com/JayjeetAtGithub/iox_observe_bench/blob/main/docs/oom_kill.md#reproducing-the-issue). 

### Expected behavior

`SortPreservingMerge` on streams of record batches with high-cardinality dictionary-encoded sort keys should be memory aware and keep memory usage within a user-defined limit. 

### Additional context

**Possible solution:**

1. Keep track of the memory usage for the `RowConverter` using the `size()` method which in the case of `Dictionary` fields returns the size of the `OrderPreservingInterner`.
2. If the size of the `RowConverter` grows more than a user-defined memory limit, take note of the `RowCursorStream` that are still getting converted, delete the converter, create a new one, and re-do the aborted conversions.

### PR DESCRIPTION
**Pull request by JayjeetAtGithub**
## Which issue does this PR close?

Closes #7200.

## Rationale for this change

This PR uses the `CardinalityAwareRowConverter` implemented in `arrow-row` in [this](https://github.com/apache/arrow-rs/pull/4736) PR.

### CODE REVIEW

**Comment by alamb** [comment]
@JayjeetAtGithub  -- I think after some thought we should put this code into DataFusion (it will need to be changed).

The reason I think this belongs in DataFusion is that the arrow row converter already has a way to control if interning is used for dictionaries. I seems somewhat confusing to then also have something that automatically picks an interning strategy based on cardinality -- I think until we see others wanting to use this, let's leave this in DataFusion

File: datafusion/core/src/physical_plan/sorts/stream.rs [hunk 1]
```diff
@@ -80,7 +80,7 @@ impl FusedStreams {
 #[derive(Debug)]
 pub struct RowCursorStream {
     /// Converter to convert output of physical expressions
-    converter: RowConverter,
```
[the remark this answers is not in the document]
Comment (JayjeetAtGithub) [reply datafusion/core/src/physical_plan/sorts/stream.rs:83]: Looks like the RowConverter struct has private fields `fields` and `codecs`, and `codec::Codec` is also not public. That's why I am unable to access these from DataFusion.

[the remark this answers is not in the document]
Comment (JayjeetAtGithub) [reply datafusion/core/src/physical_plan/sorts/stream.rs:83]: @tustvold am I correct ?

[the remark this answers is not in the document]
Comment (tustvold) [reply datafusion/core/src/physical_plan/sorts/stream.rs:83]: You shouldn't need to access these private members from DataFusion, there are public constructors?

[the remark this answers is not in the document]
Comment (JayjeetAtGithub) [reply datafusion/core/src/physical_plan/sorts/stream.rs:83]: Got it, thanks.

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs [hunk 2]
```diff
@@ -286,10 +290,123 @@ mod tests {
...
-    use arrow::array::{Int32Array, StringArray, TimestampNanosecondArray};
+    use arrow::array::{Int32Array, StringArray, TimestampNanosecondArray, DictionaryArray};
+
+    use crate::physical_plan::streaming::PartitionStream;
+    use crate::physical_plan::stream::RecordBatchStreamAdapter;
+    use crate::datasource::{streaming::StreamingTable, TableProvider};
 
     use super::*;
 
+    fn make_infinite_sorted_stream(col_b_init: &u32) -> BoxStream<'static, RecordBatch> {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:301]: This doesn't appear to yield a sorted stream?

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs [hunk 3]
```diff
@@ -286,10 +290,123 @@ mod tests {
...
+                col_b_ascii = col_b_ascii + 2;
+            }
+
+            counter = counter + 1;
+            
+            // building col `a`
+            let mut values_vector: Vec<String> = Vec::new();
+            for _i in 1..=8192 {
+                // values_vector.push(rand::thread_rng().gen_range(1..=1000));
+                values_vector.push(String::from(Uuid::new_v4().to_string()));
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:320]: This is likely quite an expensive way to generate data, it will be using secure random generator. There should be examples of sampling Alphanumeric within the codebase

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs [hunk 4]
```diff
@@ -286,10 +290,123 @@ mod tests {
...
+            // converting the iterator into a futures::stream::Stream
+            Box::pin(RecordBatchStreamAdapter::new(
+                self.schema.clone(),
+                make_infinite_sorted_stream(&self.col_b_init).map(Ok)
+            ))
+        }
+    }
+
+    #[tokio::test]
+    async fn test_dict_merge_infinite() {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:368]: What is this a test of?

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs [hunk 5]
```diff
@@ -286,10 +290,123 @@ mod tests {
...
+            },
+        ];
+        
+        let provider = StreamingTable::try_new(schema, vec![stream_1, stream_2]).unwrap();
+
+        let plan = provider.scan(&session_ctx.state(), None, &[], None).await.unwrap();
+        let exec = Arc::new(SortPreservingMergeExec::new(sort, plan));
+        let mut stream = exec.execute(0, task_ctx).unwrap();
+        while let Some(batch) = stream.next().await {
+            println!("{}", arrow::util::pretty::pretty_format_batches(&[batch.unwrap().clone()])
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:404]: ?

File: datafusion/core/src/physical_plan/wrapper.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        self.inner.as_ref().unwrap().convert_rows(rows)
+    }
+
+    pub fn convert_columns(
+        &mut self,
+        columns: &[ArrayRef]) -> Result<Rows, ArrowError> {
+        
+        if !self.done {
+            for (i, col) in columns.iter().enumerate() {
+                if let DataType::Dictionary(k, _) = col.data_type() {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/wrapper.rs:65]: This could make use of https://docs.rs/arrow-array/latest/arrow_array/cast/trait.AsArray.html#tymethod.as_any_dictionary_opt to avoid the downcasting

File: datafusion/core/src/physical_plan/wrapper.rs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+pub struct CardinalityAwareRowConverter {
+    fields: Vec<SortField>,
+    inner: Option<RowConverter>,
+    done: bool,
+}
+
+impl CardinalityAwareRowConverter {
+    pub fn new(fields: Vec<SortField>) -> Result<Self, ArrowError> {
+        Ok(Self {
+            fields: fields.clone(),
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/wrapper.rs:45]: This clone appears to be redundant

File: datafusion/core/src/physical_plan/wrapper.rs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+    pub fn convert_rows(&self, rows: &Rows) -> Result<Vec<ArrayRef>, ArrowError> {
+        self.inner.as_ref().unwrap().convert_rows(rows)
+    }
+
+    pub fn convert_columns(
+        &mut self,
+        columns: &[ArrayRef]) -> Result<Rows, ArrowError> {
+        
+        if !self.done {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/wrapper.rs:63]: If you made `fields: Option<Vec<SortField>>` you could drop self.done and this would become `if let Some(fields) = self.fields`

Alternatively an explicit enum might be better


```
pub struct CardinalityAwareRowConverter(ConverterState);

enum ConverterState {
    Init(Vec<SortField>),
    Ready(RowConverter)
}

```

File: datafusion/core/src/physical_plan/mod.rs [hunk 9]
```diff
@@ -402,6 +402,7 @@ pub mod union;
 pub mod unnest;
 pub mod values;
 pub mod windows;
+pub mod wrapper;
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/mod.rs:405]:
```suggestion
mod wrapper;
```

I'm not sure this needs to be public

**Review by tustvold — COMMENTED** [review]
Left some comments, I think it would also be good to check the benchmarks to see how this impacts performance

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs (hunk 4 above)
Comment (JayjeetAtGithub) [reply datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:368]: This will be removed. For now, this is just a rough test to check if memory is going out of bounds.

File: datafusion/core/src/physical_plan/wrapper.rs (hunk 8 above)
Comment (JayjeetAtGithub) [reply datafusion/core/src/physical_plan/wrapper.rs:50]: Umm , as per my understanding,`fields` is never None. We initialize `CardinalityAwareRowConverter` with some fields, and then modify it according to the first batch.

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs (hunk 4 above)
Comment (JayjeetAtGithub) [reply datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:368]: I would work on adding a test to check if the `CardinalityAwareRowConverter` can switch the `preserve_dictionaries` according to the row cardinality. But do you have any suggestion to put the current e2e test to check the memory usage ?

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs (hunk 5 above)
Comment (JayjeetAtGithub) [reply datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:398]: Will remove soon !

File: datafusion/core/src/physical_plan/wrapper.rs (hunk 8 above)
Comment (JayjeetAtGithub) [reply datafusion/core/src/physical_plan/wrapper.rs]: Thanks for the pointer. Pushed the change.

**Comment by JayjeetAtGithub** [comment]
> Left some comments, I think it would also be good to check the benchmarks to see how this impacts performance

@tustvold I benchmarked the current implementation of `CardinalityAwareRowConverter` with `RowConverter` and didn't see any performance regression. See the results [here](https://github.com/JayjeetAtGithub/iox_observe_bench/blob/main/docs/bench.md#benchmarking-row-conversion-on-low-and-high-cardinality-dicts-w-dict-preserving).

File: datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs (hunk 4 above)
Comment (JayjeetAtGithub) [reply datafusion/core/src/physical_plan/sorts/sort_preserving_merge.rs:363]: For now, I have added `#[ignore]` to this test, so that it does not get run in CI

**Comment by alamb** [comment]
I plan to review this PR shortly

File: datafusion/core/src/physical_plan/wrapper.rs [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+    pub fn convert_columns(
+        &mut self,
+        columns: &[ArrayRef]) -> Result<Rows, ArrowError> {
+        if self.fields != None {
+            let mut updated_fields = self.fields.take();
+            for (i, col) in columns.iter().enumerate() {
+                if let DataType::Dictionary(_, _) = col.data_type() {
+                    let cardinality = col.as_any_dictionary_opt().unwrap().values().len();
+                    if cardinality >= LOW_CARDINALITY_THRESHOLD {
+                        updated_fields.as_mut().unwrap()[i] = updated_fields.as_ref().unwrap()[i].clone().preserve_dictionaries(false);
```
Comment (alamb) [inline datafusion/core/src/physical_plan/wrapper.rs:57]: You could probably avoid this `as_mut` and `unwrap` by creating the new fields in parallel, something like

```rust
 let updated_fields: Vec<_> = updated_fields.into_iter().zip(columns.into_iter())
                .map(|(field, col)| {
...
```

File: datafusion/core/src/physical_plan/wrapper.rs [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    async fn test_with_high_card() {
+        let batch = generate_batch_with_cardinality(String::from("high")).unwrap();
+        let sort_fields = vec![
+            arrow::row::SortField::new_with_options(DataType::Dictionary(Box::new(DataType::Int32), Box::new(DataType::Utf8)), SortOptions::default()),
+            arrow::row::SortField::new_with_options(DataType::Int32, SortOptions::default())
+        ];
+        let mut converter = CardinalityAwareRowConverter::new(sort_fields.clone()).unwrap();
+        let rows = converter.convert_columns(&batch.columns()).unwrap();        
+        let converted_batch = converter.convert_rows(&rows).unwrap();
+        assert_eq!(converted_batch[0].data_type(), &DataType::Utf8);
```
Comment (alamb) [inline datafusion/core/src/physical_plan/wrapper.rs:135]: it might be worth a comment here explaining that we expect the type to be changed

File: datafusion/core/src/physical_plan/wrapper.rs [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+                    }
+                }
+            }
+            self.inner = Some(RowConverter::new(updated_fields.unwrap())?);
+        }
+        self.inner.as_mut().unwrap().convert_columns(columns)
+    }
+}
+
+mod tests {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/wrapper.rs:67]: Typically I think tests have the `#[cfg(test)]` guard on them:

```suggestion
#[cfg(test)]
mod tests {
```

**Review by alamb — COMMENTED** [review]
Thanks @JayjeetAtGithub  -- this PR is looking close. I left a few comments and there appears to be a CI test failure and clippy issues that need to be resolved. 

Also, is there an end to end test you can add showing how this PR avoids a memory explosion while mergning?

**Comment by JayjeetAtGithub** [comment]
@alamb I am working on testing out the wrapper in IOx over the OOM'ed Jaegar queries and adding some changes necessary to make that work.

**Review by alamb — COMMENTED** [review]
Looking better all the time!

I pushed some comments to this PR and ran `cargo fmt` to get some of the lints to pass

**Comment by alamb** [comment]
I am going to work on this PR a little now and try and get the tests to pass

**Comment by JayjeetAtGithub** [comment]
@alamb Thanks a lot for taking a look. Some thoughts/observations regarding the failing tests:

Before exec: `dict<int, utf8>, int` -> `CardinalityAwareRowConvert::convert_columns` (preserve_dict=false) -> `Row`
After exec: `Row`-> `CardinalityAwareRowConvert::convert_rows` -> `utf8, int`

So, basically the exec node returns `utf8` instead of `dict`, which is what we want, but it causes type mismatch for the next exec nodes in the query plan, which is what we see in the test failures. Since, I was working on fixing the `GROUP BY`, I added code to convert back `utf8` arrays into `dict` arrays in the `GroupedHashAggregateStream`.  I think if we do something similar for the other exec nodes that deal with the cardinality aware row converter, we should be good. 

Unfortunately, all these does not yet fix the `GROUP BY` OOM issue yet, and I am focusing on reproducing and fixing that end-to-end and verifying if the conversion back to `dict` is not causing any issues.

**Comment by alamb** [comment]
> Unfortunately, all these does not yet fix the GROUP BY OOM issue yet, and I am focusing on reproducing and fixing that end-to-end and verifying if the conversion back to dict is not causing any issues.

I think you may find that https://github.com/apache/arrow-datafusion/pull/7192 will reduce the memory required for the group by query significantly

I don't expect that this PR will reduce the requirements for that much

**Comment by alamb** [comment]
>  Since, I was working on fixing the GROUP BY, I added code to convert back utf8 arrays into dict arrays in the GroupedHashAggregateStream. I think if we do something similar for the other exec nodes that deal with the cardinality aware row converter, we should be good.

@JayjeetAtGithub 

I think we should try and move this logic into `CardinalityAwareRowConverter` itself -- that way we don't have to replicate the logic in multiple exec nodes (and any future users will get it for free). I also think @tustvold  has some plan to do the reconversion directly in the `RowConverter` -- I will work on sorting out / finding a ticket reference tomorrow

**Comment by JayjeetAtGithub** [comment]
@alamb @tustvold I tried my hands in moving the reconversion to dict logic inside the `RowConverter` in this [PR](https://github.com/JayjeetAtGithub/arrow-datafusion/pull/3). It is probably not the the best way, but I anyways gave it a shot. Since, it has some refactoring, I decided not to push it directly in this branch.

**Comment by alamb** [comment]
> @alamb @tustvold I tried my hands in moving the reconversion to dict logic inside the RowConverter in this https://github.com/JayjeetAtGithub/arrow-datafusion/pull/3. It is probably not the the best way, but I anyways gave it a shot. Since, it has some refactoring, I decided not to push it directly in this branch.

Thanks @JayjeetAtGithub  -- I'll incorporate it into this PR

**Comment by alamb** [comment]
I think this PR is almost ready to go -- the only thing I know that is left to do is to update the sort memory test now that less memory is used in merging. I know what to do and will fix it shortly

CI is also intermittently failing with this error: https://github.com/apache/arrow-datafusion/issues/7523 which I would like to sort out sooner rather than later

**Comment by JayjeetAtGithub** [comment]
> I think this PR is almost ready to go -- the only thing I know that is left to do is to update the sort memory test now that less memory is used in merging. I know what to do and will fix it shortly

Thanks a lot @alamb , This PR is looking great now !

**Comment by alamb** [comment]
Ok, I think this is now ready for some additional review.

File: datafusion/core/src/physical_plan/row_converter.rs [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+                output_types,
+            } => {
+                // Cast output type if needed
+                let output = converter
+                    .convert_rows(rows)?
+                    .into_iter()
+                    .zip(output_types.iter())
+                    .map(|(arr, output_type)| {
+                        if arr.data_type() != output_type {
+                            Ok(arrow::compute::cast(&arr, output_type)?)
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/row_converter.rs:125]: It is perhaps worth highlighting that if the dictionary is indeed high cardinality, this conversion will be quite expensive. I don't have a good solution for this

File: datafusion/core/src/physical_plan/row_converter.rs [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+    /// initializing the converter based on cardinalities
+    pub fn convert_columns(&mut self, columns: &[ArrayRef]) -> Result<Rows> {
+        Ok(self.converter_mut(columns)?.convert_columns(columns)?)
+    }
+
+    /// Return a mutable reference to the inner converter, creating it if needed
+    fn converter_mut(&mut self, columns: &[ArrayRef]) -> Result<&mut RowConverter> {
+        if let Self::New { fields } = self {
+            // TODO clean up the code
+            let mut updated_fields = fields.clone();
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/row_converter.rs:147]: Could use std::mem::take here to avoid a clone

File: datafusion/core/src/physical_plan/row_converter.rs [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+            let mut output_types = vec![];
+            for (i, col) in columns.iter().enumerate() {
+                output_types.push(col.data_type().clone());
+                if let DataType::Dictionary(_, _) = col.data_type() {
+                    // see comments on LOW_CARDINALITY_THRESHOLD for
+                    // the rationale of this calculation
+                    let cardinality = col.as_any_dictionary_opt().unwrap().values().len();
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/row_converter.rs:154]: Nice to see this getting some usage :heart:

File: datafusion/core/tests/memory_limit.rs [hunk 16]
```diff
@@ -238,7 +238,7 @@ async fn sort_preserving_merge() {
 
 #[tokio::test]
 async fn sort_spill_reservation() {
-    let partition_size = batches_byte_size(&dict_batches());
+    let partition_size = batches_byte_size(&maybe_split_batches(dict_batches(), true));
```
Comment (tustvold) [inline datafusion/core/tests/memory_limit.rs:241]: What is the rationale for this change?

File: datafusion/core/src/physical_plan/row_converter.rs [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+/// (aka the length of the values array) is a more robust predictor of
+/// being "high" cardinality than the actual number of keys used. The
+/// intuition for this is that other values in the dictionary could be
+/// used in subsequent batches.
+///
+/// While an argument can made for doing something more sophisticated,
+/// this would likely only really make sense if the dictionary
+/// interner itself followed a similar approach, which it did not at
+/// the time of this writing.
+const LOW_CARDINALITY_THRESHOLD: usize = 10;
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/row_converter.rs:45]: I'm curious if this has been determined empirically, a dictionary will almost never have this few values, even if the column itself only contains 2 or 3 distinct values. This might be a very strong argument for removing the dictionary interning from RowConverter...

Comment (alamb) [reply datafusion/core/src/physical_plan/row_converter.rs:45]: > a dictionary will almost never have this few values, even if the column itself only contains 2 or 3 distinct values. 

I don't understand that statement. 

@JayjeetAtGithub do you have any specific numbers about how this threshold was chosen? @tustvold do you have any suggestions for a different value (or suggestions on how to empirically determine a threshold?)

Comment (tustvold) [reply datafusion/core/src/physical_plan/row_converter.rs:45]: > I don't understand that statement.

Even if an array has very few logically distinct values, there are various mechanisms that will lead the dictionary to contain duplicates and/or values that aren't referenced. As a result almost all dictionaries, certainly any that originate from parquet files, are likely to contain more than 10 values.

> do you have any suggestions for a different value (or suggestions on how to empirically determine a threshold?)

I mean https://github.com/apache/arrow-rs/issues/4811 effectively proposes a value of 0... :sweat_smile:

File: datafusion/core/src/physical_plan/row_converter.rs (hunk 13 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/row_converter.rs:125]: One potential solution is to move it upstream in arrow-rs and avoid the intermediate copy, perhaps. I have filed a ticket https://github.com/apache/arrow-rs/issues/4813 to track this idea

File: datafusion/core/src/physical_plan/row_converter.rs (hunk 17 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/row_converter.rs:45]: Yeah, the more I think about the costs and benefits of dictionary encoding, the more I am leaning towards trying to remove dictionary encoding entirely in the row format. 

https://github.com/apache/arrow-rs/issues/4811#issuecomment-1717787949

File: datafusion/core/src/physical_plan/row_converter.rs [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+    /// Calls [`RowConverter::convert_columns`] after first
+    /// initializing the converter based on cardinalities
+    pub fn convert_columns(&mut self, columns: &[ArrayRef]) -> Result<Rows> {
+        Ok(self.converter_mut(columns)?.convert_columns(columns)?)
+    }
+
+    /// Return a mutable reference to the inner converter, creating it if needed
+    fn converter_mut(&mut self, columns: &[ArrayRef]) -> Result<&mut RowConverter> {
+        if let Self::New { fields } = self {
+            // TODO clean up the code
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/physical_plan/row_converter.rs]: I think it is a left over -- I will remove it

File: datafusion/core/src/physical_plan/row_converter.rs (hunk 14 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/row_converter.rs:150]: Sadly it seems like `SortOptions` does not implement `Default` 😢 


```
   --> datafusion/core/src/physical_plan/row_converter.rs:150:53
    |
150 |             let mut updated_fields = std::mem::take(&mut fields);
    |                                      -------------- ^^^^^^^^^^^ the trait `std::default::Default` is not implemented for `&mut std::vec::Vec<arrow::arrow_row::SortField>`
    |                                      |
    |                                      required by a bound introduced by this call
    |
note: required by a bound in `std::mem::take`
```

File: datafusion/core/tests/memory_limit.rs (hunk 16 above)
Comment (alamb) [reply datafusion/core/tests/memory_limit.rs]: it is needed to match the change in data (which is also split). I tried to make this clearer in 355ef7397.

File: datafusion/core/src/physical_plan/row_converter.rs [hunk 19]
```diff
@@ -0,0 +1,292 @@
...
+    {
+        match self {
+            Self::New { .. } => internal_err!(
+                "CardinalityAwareRowConverter has not converted any rows yet"
+            ),
+            Self::Converting {
+                converter,
+                output_types,
+            } => {
+                // Cast output type if needed
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/core/src/physical_plan/row_converter.rs]: in b673c09b0

**Comment by alamb** [comment]
Update: I think this PR is ready to go except for figuring out what the proper value for the LOW_CARDINALITY cutoff is

I believe @tustvold  is checking setting it to zero via https://github.com/apache/arrow-rs/issues/4811. @JayjeetAtGithub  can you look into seeing what the performance threshold is? 

The idea would be to test the performance of merge on a column of different cardinalities -- maybe cardinality `4`, `8`, `12`, `20`, `50` and `100`. Maybe there is an existing benchmark that could be used in https://github.com/apache/arrow-rs/blob/master/arrow/benches/row_format.rs

**Comment by JayjeetAtGithub** [comment]
Sure, I can do that

**Comment by JayjeetAtGithub** [comment]
> The idea would be to test the performance of merge on a column of different cardinalities -- maybe cardinality 4, 8, 12, 20, 50 and 100

chart

Row conversion duration vs cardinality for dict preserving on/off. The absolute numbers are in microseconds. See [sheet](https://docs.google.com/spreadsheets/d/1ELfJaLx_VydYS_K2CSkvs1dbn2FaenpYpXxiSHjvI8M/edit?usp=sharing).

This chart shows the durations (in microseconds) taken to convert a `RecordBatch` (`dict<int,utf8>`, `int`) to the `Row` format. We sweep across the cardinality of the dictionary encoded field from `1` to `500000` and turn dictionary preserving on/off. We just measure the time for `RowConverter::convert_columns`.  See [b05919f](https://github.com/JayjeetAtGithub/arrow-datafusion/commit/b05919fb0b8da95b956d4270fe33b3ec921dc6ed).

**Comment by alamb** [comment]
I am going to resolve the merge conflicts and get this PR ready to go

**Comment by alamb** [comment]
@tustvold  and I talked about this PR -- it is likely that when arrow-rs 47 is released (with https://github.com/apache/arrow-rs/pull/4819) this code will become outdated, we think it is an improvement over the current main. Thus I plan to merge this PR when the tests pass
